pkg/tokenprovider.go

```diff
@@ -171,17 +171,57 @@ func (p TokenProvider) RefreshToken(ctx context.Context) (*TokenInfo, error) {
 
 }
 
-type Mock struct {
-	Creds
-	*MockTokenProviderImpl
-}
+func (p TokenProvider) IntrospectToken(ctx context.Context) (oidc.IntrospectionResponse, error) {
+	logging.FromContext(ctx).Debugf("Introspecting token for %s", p.creds.Endpoint())
 
-func NewMockTokenProvider(ctrl *gomock.Controller) (TokenProviderFactory, *Mock) {
-	mock := &Mock{
-		MockTokenProviderImpl: NewMockTokenProviderImpl(ctrl),
+	p.cloud.Lock()
+	if p.cloud.AccessToken == "" {
```

Comment:
Rather than having this weird lock/unlock behaviour around the call `p.AccessToken(ctx)` it would be better to add the `p.cloud.AccessToken == ""` check inside `p.AccessToken(ctx)`.

The way it's written now you can still have race conditions between two two locks. Better to lock just once.